### PR TITLE
fix(execution): Ensure that synthetic stages respect execution strategy

### PR DESCRIPTION
**Pull request by marchello2000**
Synthetic stages don't (always) get the parent context. Since the user
might set execution options (e.g. whether to fail or not the pipeline) on
the parent stage, those directives should be respected by synthetic stages.
With this change, recursively look up these directives on synthetic stages'
parent if they aren't explicitly passed down.

An example here is the deploy stages: if a synthetic generated by one of the
deploy strategies fails but the deploy stage specified to ignore failure, the
ignore behavior will not be respected.

### CODE REVIEW

File: orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt [hunk 1]
```diff
@@ -103,34 +103,47 @@ fun Stage.allAfterStagesSuccessful(): Boolean =
...
+
+      stage = if (stage.syntheticStageOwner != null) stage.parent() else null
+    }
 
-fun Stage.shouldContinueOnFailure(): Boolean =
-  context["continuePipeline"] == true
+    return default
+  }
+
+  // true if explicitly set to true
```
Comment (ajordens) [inline orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt]: Not sure this inline comment is necessary here?

File: orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt [hunk 2]
```diff
@@ -103,34 +103,47 @@ fun Stage.allAfterStagesSuccessful(): Boolean =
 fun Stage.anyBeforeStagesFailed(): Boolean =
   beforeStages().any { it.status in listOf(TERMINAL, STOPPED, CANCELED) }
 
-fun Stage.anyAfterStagesFailed(): Boolean =
```
Comment (ajordens) [inline orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt]: Why was this removed?

File: orca-kotlin/src/test/kotlin/com/netflix/spinnaker/orca/ext/StageTest.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  describe("parent stage context is empty but child is set") {
+    val pipeline = pipeline {
+      application = "test"
+      stage {
+        refId = "1"
+        type = "parent"
+        status = ExecutionStatus.SUCCEEDED
+        context = mapOf(
+          "failPipeline" to false,
+          "continuePipeline" to true)
```
Comment (ajordens) [inline orca-kotlin/src/test/kotlin/com/netflix/spinnaker/orca/ext/StageTest.kt]: conventions would say `)` on next line

File: orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt [hunk 4]
```diff
@@ -103,34 +103,47 @@ fun Stage.allAfterStagesSuccessful(): Boolean =
...
-  firstAfterStages().isNotEmpty()
-
 inline fun <reified O> Stage.mapTo(pointer: String): O = mapTo(pointer, O::class.java)
 
 inline fun <reified O> Stage.mapTo(): O = mapTo(O::class.java)
 
-fun Stage.shouldFailPipeline(): Boolean =
-  context["failPipeline"] in listOf(null, true)
+fun Stage.failureStatus(default: ExecutionStatus = TERMINAL): ExecutionStatus {
+  // Because not all stages pass the full context to their child
```
Comment (ajordens) [inline orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt]:
```
Look for `continuePipeline` or `failPipeline` attributes on parent stages in the event
that they were not explicitly passed down as part of the synthetic stage context.
```

File: orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt (hunk 2 above)
Comment (marchello2000) [reply orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt]: it wasn't used, so i figured I'd remove it

File: orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt [hunk 5]
```diff
@@ -103,34 +103,47 @@ fun Stage.allAfterStagesSuccessful(): Boolean =
...
-fun Stage.anyAfterStagesFailed(): Boolean =
-  afterStages().any { it.status in listOf(TERMINAL, STOPPED, CANCELED) }
-
 fun Stage.allAfterStagesComplete(): Boolean =
   afterStages().all { it.status.isComplete }
 
 fun Stage.hasTasks(): Boolean =
   tasks.isNotEmpty()
 
-fun Stage.hasAfterStages(): Boolean =
```
Comment (marchello2000) [inline orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt]: wasn't used

File: orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt (hunk 4 above)
Comment (marchello2000) [reply orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt]: thanks, i like this wording better

File: orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt (hunk 1 above)
Comment (marchello2000) [reply orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt]: yeah, i wanted to make clear why one is defaulted to `true` and one to `false`, but I renamed the `default` param and I think it's clearer now and no longer needs the comment

File: orca-kotlin/src/test/kotlin/com/netflix/spinnaker/orca/ext/StageTest.kt (hunk 3 above)
Comment (marchello2000) [reply orca-kotlin/src/test/kotlin/com/netflix/spinnaker/orca/ext/StageTest.kt]: 👍

**Comment by robfletcher** [comment]
I think this will work as intended but I'm not sure it's the correct behavior. For example if we have a deploy stage that should continue on failure, that doesn't mean all its synthetic stages should do the same. For example, if the disable previous server group step fails we should probably not carry on to try to destroy it.

**Comment by marchello2000** [comment]
Discussed offline, since this behavior is what happens in most cases today already (since most synthetic stages inherit the context of the parent) we are going to go forward with this behavior though, as @robfletcher points out, it's not 100% correct. We will be addressing the correctness separately.

File: orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt (hunk 2 above)
Comment (marchello2000) [reply orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt]: will put back in

**Comment by marchello2000** [comment]
closing since #2686 is a better solution
